src/rgw/rgw_compression.cc

```diff
@@ -0,0 +1,173 @@
+          newbl.old_ofs = ofs;
+          newbl.new_ofs = bs > 0 ? blocks[bs-1].len + blocks[bs-1].new_ofs : 0;
+          newbl.len = in_bl.length();
+          blocks.push_back(newbl);
+        }
+      }
+    } else {
+      compressed = false;
+      in_bl.claim(bl);
+    }
+    // end of compression stuff
+  }
+
+  return next->handle_data(in_bl, ofs, phandle, pobj, again);
+}
+
+//----------------RGWGetObj_Decompress---------------------
+RGWGetObj_Decompress::RGWGetObj_Decompress(CephContext* cct_, 
+                                           RGWCompressionInfo* cs_info_, 
+                                           bool partial_content_,
+                                           RGWGetDataCB* next): RGWGetObj_Filter(next),
+                                                                cct(cct_),
+                                                                cs_info(cs_info_),
+                                                                partial_content(partial_content_),
+                                                                q_ofs(0),
+                                                                q_len(0),
+                                                                first_data(true),
+                                                                cur_ofs(0)
+{
+  compressor = Compressor::create(cct, cs_info->compression_type);
+  if (!compressor.get())
+    lderr(cct) << "Cannot load compressor of type " << cs_info->compression_type 
+                     << " for rgw, check rgw_compression_type config option" << dendl;
+}
+
+int RGWGetObj_Decompress::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  ldout(cct, 10) << "Compression for rgw is enabled, decompress part " << bl_len << dendl;
+
+  if (!compressor.get()) {
+    // if compressor isn't available - error, because cannot return decompressed data?
+    lderr(cct) << "Cannot load compressor of type " << cs_info->compression_type 
+                     << " for rgw, check rgw_compression_type config option" << dendl;
+    return -EIO;
+  }
+  bufferlist out_bl, in_bl;
+  bl_ofs = 0;
+  if (waiting.length() != 0) {
+    in_bl.append(waiting);
+    in_bl.append(bl);        
+    waiting.clear();
+  } else {
+    in_bl.claim(bl);
+  }
+  bl_len = in_bl.length();
+  
+  while (first_block <= last_block) {
+    bufferlist tmp, tmp_out;
+    int ofs_in_bl = first_block->new_ofs - cur_ofs;
+    if (ofs_in_bl + first_block->len > bl_len) {
+      // not complete block, put it to waiting
+      int tail = bl_len - ofs_in_bl;
+      in_bl.copy(ofs_in_bl, tail, waiting);
+      cur_ofs -= tail;
+      break;
+    }
+    in_bl.copy(ofs_in_bl, first_block->len, tmp);
+    int cr = compressor->decompress(tmp, tmp_out);
+    if (cr < 0) {
+      lderr(cct) << "Compression failed with exit code " << cr << dendl;
+      return cr;
+    }
+    if (first_block == last_block && partial_content)
+      out_bl.append(tmp_out.c_str(), q_len);
```

Comment:
should avoid `bufferlist::c_str()` here - if the bufferlist has more than one `buffer::ptr`, `c_str()` has to copy them all into a single `buffer::ptr` so it can return a raw pointer

i think this is equivalent: `tmp_out.copy(0, q_len, out_bl);`